wsmaster/che-core-api-devfile/src/test/java/org/eclipse/che/api/devfile/server/DevfileIntegrityValidatorTest.java

```diff
@@ -0,0 +1,144 @@
+  private Devfile initialDevfile;
+
+  private DevfileIntegrityValidator integrityValidator;
+
+  @BeforeClass
+  public void setUp() throws Exception {
+    integrityValidator = new DevfileIntegrityValidator();
+    String devFileYamlContent =
+        Files.readFile(getClass().getClassLoader().getResourceAsStream("devfile.yaml"));
+    initialDevfile = objectMapper.readValue(devFileYamlContent, Devfile.class);
+  }
+
+  @Test
+  public void shouldValidateCorrectDevfile() throws Exception {
+    // when
+    integrityValidator.validateDevfile(initialDevfile);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp = "Duplicate tool name found:'.*'")
+  public void shouldThrowExceptionOnDuplicateToolName() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getTools().add(new Tool().withName(initialDevfile.getTools().get(0).getName()));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp = "Multiple editor tools found: '.*', '.*'")
+  public void shouldThrowExceptionOnMultipleEditors() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getTools().add(new Tool().withName("editor-2").withType("cheEditor"));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp = "Duplicate command name found:'.*'")
+  public void shouldThrowExceptionOnDuplicateCommandName() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getCommands().add(new Command().withName(initialDevfile.getCommands().get(0).getName()));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp =
+          "Found actions which refers to non-existing tools in command '.*':'no_such_tool'")
+  public void shouldThrowExceptionOnUnexistingCommandActionTool() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getCommands().get(0).getActions().add(new Action().withTool("no_such_tool"));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp = "Duplicate project name found:'.*'")
+  public void shouldThrowExceptionOnDuplicateProjectName() throws Exception {
+    Devfile broken = copyOf(initialDevfile);
+    broken.getProjects().add(new Project().withName(initialDevfile.getProjects().get(0).getName()));
+    // when
+    integrityValidator.validateDevfile(broken);
+  }
+
+  @Test(
+      expectedExceptions = DevfileFormatException.class,
+      expectedExceptionsMessageRegExp =
+          "Invalid project name found:'.*'. Name must contain only Latin letters,\\n"
```

Comment:
I think if it is possible it is better to avoid using pattern during checking message because it's a place where a bug may appear. Like method may thrown an exception with empty project name or even not project name. So, consider putting real project name instead of `.*`

Reply:
Chenged